lambdaclass/mirra_backend
# Unit scaling

Closes https://github.com/lambdaclass/champions_of_mirra/issues/311

Implements unit scaling based on level, tier and rank. Formulas are taken from [design docs](https://www.notion.so/lambdaclass/Stat-Scaling-561b06b0abc942658d117b8d96272064?pvs=4). We are fine with hardcoding them for now.

# To test

You can take a look at the new test (or champions readme) to see how to redefine a unit's level, tier and rank. Play with these values and check that the formulas change accordingly!

Also you can try running battles (check battle.md) using same-character (Muflus) units with different levels/tiers, for example:
```
{:ok, user_1} = Champions.Users.register("User1")
{:ok, user_2} = Champions.Users.register("User2")
team1 = Enum.map([1], fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Muflus").id, level: 100, tier: 1, rank: 1, selected: true, user_id: user_1.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(character: [:basic_skill, :ultimate_skill]) end)
team2 = Enum.map(1..6, fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Muflus").id, level: 1, tier: 1, rank: 1, selected: true, user_id: user_2.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(character: [:basic_skill, :ultimate_skill]) end)
Champions.Battle.Simulator.run_battle team1, team2
```
Team 1 will win this fight everytime even though it's made of only one Muflus (team 2 has 6), because team 1's unit has a level of 100 while all the other ones are set to 1.
